src/reprepro_updater/helpers.py

```diff
@@ -142,11 +159,38 @@ def invalidate_dependent(repo_dir, distro, arch, package):
     """
     Remove all dependents of the package with the same arch.
 
-    This is only valid for binary packages.
+    This is only valid for binary packages and assumes all
+    packages are in the `main` component.
     """
-    # storage for recursion
-    processed_packages = []
-    return _invalidate_dependent(repo_dir, distro, arch, package, processed_packages)
+    # Use internal repository information parser to get reverse dependents.
+    # Using reprepro for each package was resulting in huge runtimes causing timeouts
+    # on the buildfarm when invalidating low-level packages.
+    repo_info = RepositoryInfo(repo_dir, distro, arch)
+
+    # queue of dependents to iterate
+    dependents_to_process = repo_info.get_rdepends(package)
+    # the complete list of transitive dependents
+    transitive_dependents = dependents_to_process.copy()
+
+    queue_max_size = len(dependents_to_process)
+    processed_packages = 0
+    iter_times = []
+    while len(dependents_to_process) > 0:
+        _start = time.time()
+        dep = dependents_to_process.pop()
```

Comment:
This processes the pending items in a random order. In case something goes wrong it will be almost impossible to reproduce the problem.

Reply:
Do you have a suggestion about what to do differently? Using a list instead would make every unique addition linear on the size of the queue.

While I expect the set ordering to be arbitrary I would expect it to be deterministic on constant inputs so this does not concern me much.

Reply:
> Do you have a suggestion about what to do differently? Using a list instead would make every unique addition linear on the size of the queue.

You would need a list, yes. Every `append` is certainly not `O(n)` but only `O(1)` (see https://wiki.python.org/moin/TimeComplexity). Of course the sorting will take some time but I don't expect it to be significant and it would pay of in case of issues.

> While I expect the set ordering to be arbitrary I would expect it to be deterministic on constant inputs so this does not concern me much.

I don't think so. That would make it deterministic for the same input which it is not.

Reply:
> Every append is certainly not O(n) but only O(1) (see https://wiki.python.org/moin/TimeComplexity

Yes, but if we naively append every new element we'll process packages that appear multiple times in the dependency tree... multiple times. 

> That would make it deterministic for the same input which it is not.

You're right. It's deterministic within an ipython context but not in general.